openml/datasets/functions.py

```diff
@@ -169,29 +171,32 @@ def list_datasets(offset=None, size=None, status=None, **kwargs):
         If qualities are calculated for the dataset, some of
         these are also returned.
     """
-    api_call = "data/list"
-    if offset is not None:
-        api_call += "/offset/%d" % int(offset)
 
-    if size is not None:
-        api_call += "/limit/%d" % int(size)
+    return openml.utils.list_all(_list_datasets, offset=offset, size=size, status=status, **kwargs)
+
+def _list_datasets(**kwargs):
+
+    """
+    Perform api call to return a list of all datasets.
+
+    Returns
+    -------
+    datasets : dict of dicts
+    """
 
-    if status is not None:
-        api_call += "/status/%s" %status
+    api_call = "data/list"
 
     if kwargs is not None:
         for filter, value in kwargs.items():
             api_call += "/%s/%s" % (filter, value)
+    return __list_datasets(api_call)
 
-    return _list_datasets(api_call)
-
-
-def _list_datasets(api_call):
+def __list_datasets(api_call):
     # TODO add proper error handling here!
     try:
         xml_string = _perform_api_call(api_call)
-    except OpenMLServerNoResult:
-        return dict()
+    except OpenMLServerNoResult as e:
```

Comment:
Is there a reason why you changed this? I think returning a dict is more natural because it behaves like when listing an empty directory.

Reply:
Also, it appears to me that this piece of code is repeated in all listing functions. Could you please put that code into the utils module?

Reply:
There is a reason. It breaks the way that the list method in the utils works. As an example, without a limit, the previous implementation would never throw an error and we would be stuck in a loop. 

In my opinion we should not change that, as we are performing checks and also actually creating the dict there. We can however remove the try and except as we are actually catching in the utils method.

Reply:
Basically utils.list_all -> loops _list_* -> forms the api call and passes it to the __list_* function.
With the current flow it is not possible to do that.

Reply:
Okay, got it, that makes sense. Thanks for removing the try/except, this should make the code a lot easier.